daemon/list.go

```diff
@@ -86,45 +103,34 @@ func (daemon *Daemon) Containers(config *ContainersConfig) ([]*types.Container,
 
 // reduceContainer parses the user filtering and generates the list of containers to return based on a reducer.
 func (daemon *Daemon) reduceContainers(config *ContainersConfig, reducer containerReducer) ([]*types.Container, error) {
-	containers := []*types.Container{}
-
 	ctx, err := daemon.foldFilter(config)
 	if err != nil {
 		return nil, err
 	}
 
-	for _, container := range daemon.List() {
-		t, err := daemon.reducePsContainer(container, ctx, reducer)
-		if err != nil {
-			if err != errStopIteration {
-				return nil, err
-			}
-			break
-		}
-		if t != nil {
-			containers = append(containers, t)
-			ctx.idx++
-		}
-	}
-	return containers, nil
-}
+	var (
+		filtered   = daemon.containers.Filter(ctx)
+		containers = newContainerList(len(filtered))
+		wg         sync.WaitGroup
+	)
 
-// reducePsContainer is the basic representation for a container as expected by the ps command.
-func (daemon *Daemon) reducePsContainer(container *Container, ctx *listContext, reducer containerReducer) (*types.Container, error) {
-	container.Lock()
-	defer container.Unlock()
+	for _, container := range filtered {
+		wg.Add(1)
+		go func(container *Container, ctx *listContext) {
+			defer wg.Done()
+			// transform internal container struct into api structs
 
-	// filter containers to return
-	action := includeContainerInList(container, ctx)
-	switch action {
-	case excludeContainer:
-		return nil, nil
-	case stopIteration:
-		return nil, errStopIteration
+			ct, err := reducer(container, ctx)
+			if err != nil {
+				logrus.Warnf("Error inspecting container: %q, %v", container, err)
+				return
+			}
+			containers.Add(ct)
```

Comment:
I think it's better to collect everything in chan, close it after wait and then create slice in for range loop. It'll be more readable and without mutexes.